### PR TITLE
Improvements on "ansiblegate" module

### PR DESCRIPTION
**Pull request by meaksh**
### What does this PR do?
This PR adds some improvements to the "Ansiblegate" module of Salt:

- Move Ansible inventory collector from "AnsibleRoster" to Salt utils to reuse it
- Add `ansible.targets` method to gather Ansible inventory
- Add `ansible.discover_playbooks` method to help collecting playbooks.
- Fix crash when running Ansible playbooks if `ansible-playbook` CLI output is not the expected JSON.
- Fix issues when processing inventory and there are groups with no members.
- Allow new types of targets for Ansible roster

### Merge requirements satisfied?
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by meaksh** [comment]
@twangboy @Ch3LL any chance of reviewing this one? Thanks!

**Comment by meaksh** [comment]
@s0undt3ch maybe you can take a look at this one? Thanks in advance!

BTW, just wondering what is the propose of using `[:]` to get a copy of the declarated list at 
 https://github.com/saltstack/salt/pull/60056/files#diff-519146c38c0ed7bc961ffc5f408d3dd380abb8760d48e839743dbd38dc28ba15L47

File: salt/modules/ansiblegate.py [hunk 1]
```diff
@@ -365,7 +373,201 @@ def playbooks(
...
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt 'ansiblehost' ansible.targets
+        salt 'ansiblehost' ansible.targets inventory=my_custom_inventory
+
+    """
+    return __utils__["ansible.targets"](**kwargs)
```
Comment (s0undt3ch) [inline salt/modules/ansiblegate.py:406]: Do we really need to use `__utils__`?
Why can't we import and use the util function directly?
`__utils__` should have never been added to salt(personal opinion)...

File: salt/modules/ansiblegate.py [hunk 2]
```diff
@@ -365,7 +373,201 @@ def playbooks(
...
+    try:
+        retdata = json.loads(ret["stdout"])
+    except ValueError:
+        retdata = ret
     if "retcode" in ret:
         __context__["retcode"] = retdata["retcode"] = ret["retcode"]
     return retdata
+
+
+def targets(**kwargs):
```
Comment (s0undt3ch) [inline salt/modules/ansiblegate.py:385]: Why `**kwargs` and not explicit keyword arguments?

**Comment by s0undt3ch** [comment]
> @s0undt3ch maybe you can take a look at this one? Thanks in advance!
> 
> BTW, just wondering what is the propose of using `[:]` to get a copy of the declarated list at https://github.com/saltstack/salt/pull/60056/files#diff-519146c38c0ed7bc961ffc5f408d3dd380abb8760d48e839743dbd38dc28ba15L47

It's a failed implementation at separate copies of the same original list....
```python
>>> a = b = [1, 2, 3]
>>> a
[1, 2, 3]
>>> b
[1, 2, 3]
>>> a.remove(2)
>>> a
[1, 3]
>>> b
[1, 3]
>>> a = b = [1, 2, 3][:]
>>> a
[1, 2, 3]
>>> b
[1, 2, 3]
>>> a.remove(2)
>>> a
[1, 3]
>>> b
[1, 3]
>>> a = b[:] = [1, 2, 3][:]
>>> a
[1, 2, 3]
>>> b
[1, 2, 3]
>>> a.remove(2)
>>> a
[1, 3]
>>> b
[1, 2, 3]
```

ie, a bug.

:sob:

**Comment by s0undt3ch** [comment]
The correct code should

> > @s0undt3ch maybe you can take a look at this one? Thanks in advance!
> > BTW, just wondering what is the propose of using `[:]` to get a copy of the declarated list at https://github.com/saltstack/salt/pull/60056/files#diff-519146c38c0ed7bc961ffc5f408d3dd380abb8760d48e839743dbd38dc28ba15L47
> 
> It's a failed implementation at separate copies of the same original list....
> 
> ```python
> >>> a = b = [1, 2, 3]
> >>> a
> [1, 2, 3]
> >>> b
> [1, 2, 3]
> >>> a.remove(2)
> >>> a
> [1, 3]
> >>> b
> [1, 3]
> >>> a = b = [1, 2, 3][:]
> >>> a
> [1, 2, 3]
> >>> b
> [1, 2, 3]
> >>> a.remove(2)
> >>> a
> [1, 3]
> >>> b
> [1, 3]
> >>> a = b[:] = [1, 2, 3][:]
> >>> a
> [1, 2, 3]
> >>> b
> [1, 2, 3]
> >>> a.remove(2)
> >>> a
> [1, 3]
> >>> b
> [1, 2, 3]
> ```
> 
> ie, a bug.
> 
> sob

The correct code should be:
```python
>>> a = b[:] = [1, 2, 3]
>>> a
[1, 2, 3]
>>> b
[1, 2, 3]
>>> a.remove(2)
>>> a
[1, 3]
>>> b
[1, 2, 3]
>>>
```

File: salt/utils/ansible.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+log = logging.getLogger(__name__)
+
+
+def __virtual__():  # pylint: disable=expected-2-blank-lines-found-0
+    if salt.utils.path.which("ansible-inventory"):
+        return __virtualname__
+    return (False, "Install `ansible` to use inventory")
+
+
+def targets(inventory="/etc/ansible/hosts", **kwargs):
```
Comment (s0undt3ch) [inline salt/utils/ansible.py:21]: Again, why `**kwargs` and not explicit keyword arguments?

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
Here's an initial review

File: salt/modules/ansiblegate.py (hunk 1 above)
Comment (meaksh) [reply salt/modules/ansiblegate.py]: I just replaced this to use `salt.utils.ansible.targets` function directly.

File: salt/modules/ansiblegate.py (hunk 2 above)
Comment (meaksh) [reply salt/modules/ansiblegate.py]: Fixed. Now they are passed explicetely. :+1:

File: salt/utils/ansible.py (hunk 3 above)
Comment (meaksh) [reply salt/utils/ansible.py]: Fixed.

**Comment by meaksh** [comment]
@s0undt3ch thanks for the review! :+1: 

I've pushed the changes you suggested on the review

**Comment by garethgreenaway** [comment]
@meaksh Looks like this one has some merge conflicts.

**Review by garethgreenaway — DISMISSED** [review]

**Comment by meaksh** [comment]
Thanks for the review @garethgreenaway !

I've have rebased with latest `master` so conflicts are now solved.

**Review by twangboy — DISMISSED** [review]

**Comment by Ch3LL** [comment]
there is a merge conflict that needs to be resolved.

**Comment by meaksh** [comment]
@Ch3LL conflicts are now fixed :+1:

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/modules/ansiblegate.py [hunk 4]
```diff
@@ -359,13 +382,211 @@ def playbooks(
...
+    except ValueError:
+        retdata = ret
     if "retcode" in ret:
         __context__["retcode"] = retdata["retcode"] = ret["retcode"]
     return retdata
+
+
+def targets(inventory="/etc/ansible/hosts", yaml=False, export=False):
+    """
+    Return the inventory from an Ansible inventory_file
```
[the remark this answers is not in the document]
Comment (meaksh) [reply salt/modules/ansiblegate.py:406]: Let's try to make it for 3005, if possible. I'm pushing this now

File: salt/modules/ansiblegate.py [hunk 5]
```diff
@@ -365,7 +366,201 @@ def playbooks(
...
+    try:
+        retdata = json.loads(ret["stdout"])
+    except ValueError:
+        retdata = ret
     if "retcode" in ret:
         __context__["retcode"] = retdata["retcode"] = ret["retcode"]
     return retdata
+
+
+def targets(**kwargs):
```
[the remark this answers is not in the document]
Comment (meaksh) [reply salt/modules/ansiblegate.py]: This was fixed already.

**Comment by meaksh** [comment]
@Ch3LL thanks again for the review. I've addressed your comments and I think this is now ready.

I hope it can gets into 3005!

**Review by Ch3LL — DISMISSED** [review]

**Review by s0undt3ch — DISMISSED** [review]

**Comment by Ch3LL** [comment]
Unfortunately I was not able to get this into 3005 since we were on a tight timeline. I'll go ahead and update the versionadded versions to 3006 and help get this merged in. I'll also label it 3006 so it for sure makes it into that release.

**Comment by Ch3LL** [comment]
So we actually ended up rebasing master with the release, so this will actually end up in 3005 ! yay! I also updated the versionadded to 3005: https://github.com/saltstack/salt/pull/62331

**Pull request merged**